---
title: "01.scRNA-seq: fit P, NB, ZIP, ZINM models using simulated data"
description: |
  Simulate data following one of the four distritutions and estimate parameters in R.
author: Ming
date: 2021-11-21
output:
  distill::distill_article:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This is the 2nd post of the 'modelling gene expression with single cell data' series.

Before jumping into real scRNA-seq data, in this post, I'd test [the four most frequently used statistical models in single cell gene expression modelling](https://www.nature.com/articles/s41588-021-00873-4) in R.

- P Possion model  
- NB Negative Binomial model 
- ZIP Zero-Inflated Poisson
- ZINB Zero-Inflated Negative Binomial

I'd simulate some data following certain probability distribution and use functions inside R to fit model and perform parameter estimation.

## Possion model
```{r fig.height=8}
# simulate 500 numbers following Possion model 
set.seed(2049)
y <- rpois(n = 500, lambda = 4)
n <- 500
group <- sample(c(0,1), size = n, replace = TRUE)
y_sim <- rpois(n = n, lambda = exp(-2 + 0.5 * (group == 1)))
table(y_sim,group)
par(mfrow=c(2,1))
hist(y_sim[group==0],main='Possion, group0')
hist(y_sim[group==1],main='Possion, group1')
# fit model
m1 <- glm(y_sim ~ group, family = poisson(link='log') )
summary(m1)

m0 <- update(m1, . ~ 1)
logLik(m0)
logLik(m1)
pchisq(2 * (logLik(m1) - logLik(m0)), df = 1, lower.tail = FALSE)
#https://norcalbiostat.github.io/AppliedStatistics_notes/model-fit-criteria.html
AIC(m0,m1) #Smaller is better
BIC(m0,m1) #Smaller is better
```

## Zero-inflated Possion model
Ref: https://stats.idre.ucla.edu/r/dae/zip/

A Zero-inflated Possion model has two component: inlfated 0 and a Possion distribution.

Thus, the probability of observing a 0 equals:

$$ P(Y=0) = \pi + (1-\pi) *Poisson(0|lambda) $$

```{r}
#simulate data following zero-inflated Possion model 
#https://stats.stackexchange.com/questions/189005/simulate-from-a-zero-inflated-poisson-distribution/189052
set.seed(2049)
n=1000; 
pi=0.2;  #which means 1-pi=0.8 to be Poisson
lambda=2; #lambda parameter in the Possion distribution
y=ifelse(rbinom(n, size = 1, prob = 1-pi) == 0, 0, rpois(n, lambda = lambda))
group=sample(c('a','b'),n,replace = T)
simu.dat=data.frame(y=y,group=group)

par(mfrow=c(1,1))
hist(y,main='Zero-inflated Possion')

# fit model
library(pscl)
m0 <- zeroinfl(y ~ 1, data = simu.dat)
x=summary(m0)
x
x$coefficients
exp(x$coefficients$count[1]) #possion, link.func=log
exp(x$coefficients$zero[1])/(1+exp(x$coefficients$zero[1])) # zero inflation which is not due to Poisson, link.func=logit

# maximal likelihood 
m1 <- update(m0, . ~ group)
logLik(m0)
logLik(m1)
pchisq(2 * (logLik(m1) - logLik(m0)), df = 2, lower.tail = FALSE) #m1 is not sig. better than m0. as in simulation, i just randomly assign numbers to two groups.
AIC(m0,m1)
BIC(m0,m1)
```

## Negative Binomial  model

I used `rnbinom` function in `stats` R to simulate numbers.

Some [explanations](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/NegBinomial.html) about this function:

>rnbinom(n, size, prob, mu)

>size: target for number of successful trials, or dispersion parameter (the shape parameter of the gamma mixing distribution). Must be strictly positive, need not be integer.

>An alternative parametrization (often used in ecology) is by the mean mu (see above), and size, the dispersion parameter, where prob = size/(size+mu). The variance is mu + mu^2/size in this parametrization.

```{r}
# simulate negative binomial model
#https://data.library.virginia.edu/simulating-data-for-count-models/
set.seed(2049)
n <- 1000
y_sim <- rnbinom(n = n, mu =5, size = 2)

par(mfrow=c(1,1))
hist(y_sim,main='Negative Binomial model')

m3 <- MASS::glm.nb(y_sim ~ 1)
summary(m3)
# mu: (link.function=log)
# Theta: the size parameter used in simulating data
exp(m3$coefficients)
m3$theta

# add one covariate 
set.seed(2049)
n <- 500
group <- sample(c(0,1), size = n, replace = TRUE)
y_sim <- rnbinom(n = n, 
                 mu =exp(-2 + 0.7 * (group == 1)),
                 size = 0.05)
m1 <- MASS::glm.nb(y_sim ~ group)
summary(m1)
m1$coefficients
```

## Zero-inflated Negative Binomial model

A Zero-inflated Negative Binomial model has two component: inlfated 0 and a Negative Binomial distribution.

Thus, the probability of observing a 0 equals:

$$ P(Y=0) = \pi + (1-\pi) *NB(0|mu,size) $$


```{r}
# simulate zero-inflated negative binomial model
set.seed(2049)
n <- 1000
group <- sample(c(0,1), size = n, replace = TRUE)

pi=0.1; # pi=0.1, which means 1-pi=0.9 follows NB distribution
z <- rbinom(n = n, size = 1, prob = 1-pi) 
# mean(z == 0)
y_sim <- ifelse(z == 0, 0, 
                rnbinom(n = n, 
                        mu = exp(1.3 + 1.5 * (group == 1)), 
                        size = 2))

par(mfrow=c(2,1))
hist(y_sim[group==0],main='Zero-inflated Negative Binomial, group 0')
hist(y_sim[group==1],main='Zero-inflated Negative Binomial, group 1')

m1=pscl::zeroinfl(y_sim~group,dist='negbin')
m1
x1=summary(m1)
x1
# mu: (link.function=log)
# Theta: the size parameter used in simulating data

# mu in NB
x1$coefficients$count[1,1]
# theta, or size in NB
exp(x1$coefficients$count[3,1]) #size, same as theta
x1$theta
# prob of 0 in ZI not attributed to NB 
g0=x1$coefficients$zero[1] #for group 0 
g1=x1$coefficients$zero[1]+x1$coefficients$zero[2] #for group 1 
exp(g0)/(1+exp(g0)) #prob
exp(g1)/(1+exp(g1)) #prob
```

```{r}
sessionInfo()
```


